src/TokenHandler.mo

```diff
@@ -0,0 +1,364 @@
+      if (item.credit + usableDeposit < 0) {
+        Debug.trap("item.credit + Int.max(0, item.deposit - fee) < 0");
+      };
+
+      {
+        var deposit = item.deposit;
+        var credit = item.credit;
+        var usable_balance = Int.abs(item.credit + usableDeposit);
+      };
+    };
+
+    /// retrieve the current size of consolidation backlog
+    public func backlogSize() : Nat = backlog.size();
+
+    /// deduct amount from P’s usable balance. Return false if the balance is insufficient.
+    public func debit(principal : Principal, amount : Nat) : Bool {
+      map.change(
+        principal,
+        func(info) {
+          if (info.deposit + info.credit < amount) return false;
+          info.credit -= amount;
+          return true;
+        },
+      );
+    };
+
+    ///  add amount to P’s usable balance (the credit is created out of thin air)
+    public func credit(principal : Principal, amount : Nat) : () {
+      ignore map.set(
+        principal,
+        func(info) {
+          info.credit += amount;
+          true;
+        },
+      );
+    };
+
+    /// The handler will call icrc1_balance(S:P) to query the balance. It will detect if it has increased compared
+    /// to the last balance seen. If it has increased then it will adjust the deposit (and hence the usable_balance).
+    /// It will also schedule or trigger a “consolidation”, i.e. moving the newly deposited funds from S:P to S:0.
+    public func notify(principal : Principal) : async* () {
+      if (not map.lock(principal)) return;
+      try {
+        let latestBalance = await* loadBalance(principal);
+        updateDeposit(principal, latestBalance);
+        if (latestBalance > fee) {
+          // schedule consolidation for this principal
+          backlog.push(principal);
+          ignore processBacklog();
+        };
+        map.unlock(principal);
+      } catch err {
+        map.unlock(principal);
+        throw err;
+      };
+    };
+
+    /// process first account from backlog
+    public func processBacklog() : async () = async {
+      func consolidate(principal : Principal) : async () {
+        let latestBalance = await* loadBalance(principal);
+        if (latestBalance <= fee) return;
+
+        updateDeposit(principal, latestBalance);
+
+        let transferResult = try {
+          await icrc1Ledger.icrc1_transfer({
+            from_subaccount = ?toSubaccount(principal);
+            to = { owner = ownPrincipal; subaccount = null };
+            amount = latestBalance - fee;
+            fee = null;
+            memo = null;
+            created_at_time = null;
+          });
+        } catch (err) {
+          #Err(#CallIcrc1LedgerError);
+        };
+
+        switch (transferResult) {
+          case (#Ok _) {
+            updateDeposit(principal, 0);
+            credit(principal, latestBalance - fee);
+            // remove from backlog if present
+            backlog.remove(principal);
```

Comment:
Is `backlog.remove` necessary? We popped it from backlog and then obtained a lock. So nothing else could have put it back in backlog, or?